src/libs/SelectionScraper/BaseSelectionScraper.js

```diff
@@ -0,0 +1,156 @@
+            // Assuming we selected only "block" part of following html:
+            // <div className="..." style="..." data-testid="pre">
+            //     <div dir="auto" class="..." style="...">
+            //         this is block code
+            //     </div>
+            // </div>
+            // commonAncestorContainer: #text "this is block code"
+            // commonAncestorContainer.parentNode:
+            //     <div dir="auto" class="..." style="...">
+            //         this is block code
+            //     </div>
+            // and finally commonAncestorContainer.parentNode.closest('data-testid') is targeted dom.
+            if (range.commonAncestorContainer instanceof HTMLElement) {
+                node = range.commonAncestorContainer.closest(`[${tagAttribute}]`);
+            } else {
+                node = range.commonAncestorContainer.parentNode.closest(`[${tagAttribute}]`);
+            }
+
+            // This means "range.commonAncestorContainer" is a text node. We simply get its parent node.
+            if (!node) {
+                node = range.commonAncestorContainer.parentNode;
+            }
+
+            node = node.cloneNode();
+            node.appendChild(clonedSelection);
+            div.appendChild(node);
+        }
+    }
+
+    return div.innerHTML;
+};
+
+/**
+ * Clears all attributes from dom elements
+ * @param {Object} dom htmlparser2 dom representation
+ * @returns {Object} htmlparser2 dom representation
+ */
+const replaceNodes = (dom) => {
+    let domName = dom.name;
+    let domChildren;
+    const domAttribs = {};
+    let data;
+
+    // Encoding HTML chars '< >' in the text, because any HTML will be removed in stripHTML method.
+    if (dom.type === 'text') {
+        data = Str.htmlEncode(dom.data);
+    }
+
+    // We are skipping elements which has html and body in data-testid, since ExpensiMark can't parse it. Also this data
+    // has no meaning for us.
+    if (dom.attribs && dom.attribs[tagAttribute]) {
+        if (!elementsWillBeSkipped.includes(dom.attribs[tagAttribute])) {
+            domName = dom.attribs[tagAttribute];
+        }
+
+        // Adding a new line after each comment here, because adding after each range is not working for chrome.
+        if (htmlEngineUtils.isCommentTag(dom.attribs[tagAttribute])) {
+            dom.children.push(new Element('br', {}));
+        }
+    }
+
+    // We need to preserve href attribute in order to copy links.
+    if (dom.attribs && dom.attribs.href) {
+        domAttribs.href = dom.attribs.href;
+    }
+
+    if (dom.children) {
+        domChildren = _.map(dom.children, c => replaceNodes(c));
+    }
+
+    return {
+        ...dom,
+        data,
+        name: domName,
+        attribs: domAttribs,
+        children: domChildren,
+    };
+};
+
+/**
+ * Resolves the selection to values and produces clean HTML.
+ * @param {Object} selectedHtml html to parse
+ * @returns {Object} resolved HTML in the selection format
```

Comment:
```suggestion
 * @returns {?Object} resolved HTML in the selection format
```

Reply:
It returns null if no HTML is produced.

Reply:
Should I change this like you say and pass it and consequently use it as a potentially null prop, or should there always be an empty object? This is related to the above error we throw and the comments below.

Reply:
null is probably fine. It will be easier to detect.